cmd/kubelet/app/server.go

```diff
@@ -412,13 +431,249 @@ func run(s *options.KubeletServer, kcfg *KubeletConfig) (err error) {
 	return nil
 }
 
+// bootstrapClientCert will request a client cert for kubelet.
+// If the file specified by --kubeconfig does not exist, the bootstrap kubeconfig is used
+// to request a client certificate from the API server.
+// On success, a kubeconfig file referencing the generated key and obtained certificate is
+// written to the path specified by --kubeconfig.
+// The certificate and key file will be stored in /var/run/kubernetes/.
+func bootstrapClientCert(s *options.KubeletServer) error {
+	// Check the if --kubeconfig already has sufficient TLS client info.
+	kcfg, err := (&clientcmd.ClientConfigLoadingRules{ExplicitPath: s.KubeConfig.Value()}).Load()
+	if err == nil {
+		authInfo, err := getCurrentContextAuthInfo(kcfg)
+		if err == nil {
+			if containsSufficientTLSInfo(authInfo) {
+				return nil
+			}
+		}
+	}
+
+	// At this point, we need to use the bootstrap kubeconfig to generate TLS client cert, key, and a kubeconfig
+	// to stored in --kubeconfig.
+	glog.V(2).Info("Using bootstrap kubeconfig to generate TLS client cert, key and kubeconfig file")
+
+	kcfg, err = (&clientcmd.ClientConfigLoadingRules{ExplicitPath: s.BootstrapKubeconfig}).Load()
+	if err != nil {
+		return fmt.Errorf("unable to load boostrap kubeconfig: %v", err)
+	}
+	authInfo, err := getCurrentContextAuthInfo(kcfg)
+	if err != nil {
+		return fmt.Errorf("unable to load auth info in bootstrap kubeconfig: %v", err)
+	}
+
+	authInfo.ClientCertificate, authInfo.ClientKey, err = getClientCertAndKey(s, authInfo.ClientCertificate, authInfo.ClientKey)
+	if err != nil {
+		return fmt.Errorf("unable to get cert from API server: %v", err)
+	}
+
+	// Marshal and write the kubeconfig to disk.
```

Comment:
I think I would build a struct based on the restclient.Config the bootstrap kubeconfig resolved to. also, write using `clientcmd.WriteToFile` (see https://github.com/liggitt/kubernetes/blob/tls-bootstrap/cmd/kubelet/app/bootstrap.go#L104-L136 for struct building and marshaling)